### ISSUE TITLE
Add integration with JVM profiler 

### ISSUE THREAD

**Issue by WojciechMazur** [scala-native/scala-native#3597]
It would be really helpfull in terms of build times improvements, to have easy access to profiling info from within the sbt. We should try to create integration to async-profiler or other available profiler allowing for automatic collecting raports when starting `nativeLink` command (or a dedicated command which would call it) and would aggregate the results (prefered flamegraph or details with execution times) when upon finish or abort of `nativeLink` command

**Comment by tanishiking** [comment]
That sounds nice! The introduction of the [ap-loader](https://github.com/jvm-profiling-tools/ap-loader) should significantly simplify the integration of async-profiler. (it was previously [a bit cumbersome](https://mostlynerdless.de/blog/2022/11/21/ap-loader-a-new-way-to-use-and-embed-async-profiler/)  😅) 

This new loader potentially enables us to profile specific spans of the Scala-Native process by using its [Java APIs](https://github.com/jvm-profiling-tools/ap-loader#usage-in-java-code).

(Slightly off-topic: Perhaps we could extend this concept and develop an sbt plugin to streamline the profiling of sbt tasks using async-profiler in the future.)

**Comment by tanishiking** [comment]
Since [jmh has an integration with async-profiler](https://github.com/sbt/sbt-jmh#using-java-flight-recorder--async-profiler), 

Instruction:

- Download the DLL for `async-profiler` from https://github.com/async-profiler/async-profiler#download
- Enable to JMH to load `build/libasyncProfiler.so` (or `.dylib` on macos), by copying it to `java.library.path` or setting an `LD_LIBRARY_PATH`, you will see an instruction if JMH failed to load the DLL. 
  - (off topic: we can improve JMH to run async-profiler out of the box using ap-loader 😄 )
- Run benchmark with `-prof async:output=flamegraph`

```
 [error] Profilers failed to initialize, exiting.
 [error] Unable to load async-profiler. Ensure asyncProfiler library is on LD_LIBRARY_PATH (Linux), DYLD_LIBRARY_PATH (Mac OS), or -Djava.library.path. Alternatively, point to explicit library location with -prof async:libPath=<path>.
 [error] no asyncProfiler in java.library.path: /Users/tanishiking/Library/Java/Extensions:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java:.
```

```sh
sbt:scala-native> toolsBenchmarks2_13 / Jmh / run -prof async:output=flamegraph
...
[info] Benchmark                                               Mode  Cnt    Score    Error  Units
[info] LinkerBench.link                                        avgt   10  368.940 ± 40.646  ms/op
[info] LinkerBench.link:·async                                 avgt           NaN             ---
[info] OptimizeDebug.optimize                                  avgt   10  238.355 ±  1.743  ms/op
[info] OptimizeDebug.optimize:·async                           avgt           NaN             ---
[info] OptimizeReleaseFast.optimize                            avgt   10  950.020 ±  7.173  ms/op
[info] OptimizeReleaseFast.optimize:·async                     avgt           NaN             ---
[info] testinterface.CodeGen.codeGen                           avgt   10  205.641 ±  6.999  ms/op
[info] testinterface.CodeGen.codeGen:·async                    avgt           NaN             ---
[info] testinterface.CodeGenWithDebugMetadata.codeGen          avgt   10  514.271 ± 32.482  ms/op
[info] testinterface.CodeGenWithDebugMetadata.codeGen:·async   avgt           NaN             ---
[info] testinterface.CodeGenWithMultithreading.codeGen         avgt   10  399.189 ± 33.582  ms/op
[info] testinterface.CodeGenWithMultithreading.codeGen:·async  avgt           NaN             ---
```

And generates profile results in flamegraph. [profile.zip](https://github.com/scala-native/scala-native/files/13305324/profile.zip)

Do we think this is enough? While cannot profile various scenarios, profiling with benchmark suites would usually provide enough information to performance improvements.

**Comment by WojciechMazur** [comment]
That's definitely a good start and as you mentioned should help with most of case, however it would be good to also prepare a set of commands that could be applied to 3rd party projects to debug performance issues in them. These, at some point could be made an sbt plugin, but for our internal usage, some manual with changes to be applied would be good enough. 
So I'd like to go to any arbitrary project using Scala Native and add entry to build.sbt 
```scala
nativeLink := { 
  prepareProfier()
  try nativeLink.value // delegate to original implementation
  finally { 
    stopProfiler()
    createReport()
  }
}
``` 
At that point we would be able to debug compilation issue performance in the user's apps

**Comment by WojciechMazur** [comment]
Resolved in #3602
